File: go/spinner/spinner.go
package main

import (
	"fmt"
	"strings"
	"time"
)

func spin(msg string, done <-chan bool) {
	chars := []rune{'|', '/', '-', '\\'}
	i := 0
	status := ""
	Loop:
	for {
		select {
		case <- done:
			break Loop
		default:
			time.Sleep(100 * time.Millisecond)
			char := chars[i]
			i = (i + 1) % len(chars)
			status = fmt.Sprintf("%c %s", char, msg)
			fmt.Printf("%s\r", status)
		}
	}
	fmt.Printf("%s\r", strings.Repeat(" ", len(status)))
}

func slow_function() int {
    time.Sleep(3 * time.Second)
	return 42
}

func supervisor() int {
	done := make(chan bool)
	go spin("thinking!", done)
	result := slow_function()
	done <- true
	return result
}

func main() {
	result := supervisor()
	fmt.Println("Answer:", result)
}

File: elixir/spinner.exs
#! /usr/bin/env elixir

defmodule Spinner do
  # milliseconds
  @delay 100
  def spin(msg) do
    "⠇⠋⠙⠸⠴⠦"
    |> String.graphemes()
    |> Stream.cycle()
    |> spin(msg)
  end

  def spin(char_stream, msg) do
    char = Enum.at(char_stream, 0)
    status = "\r#{char} #{msg}"

    receive do
      {super_pid, :computed} ->
        blanks = String.duplicate(" ", String.length(status))
        IO.write("\r#{blanks}\r")
        send(super_pid, :done)
    after
      @delay ->
        IO.write(status)

        char_stream
        |> Stream.drop(1)
        |> spin(msg)
    end
  end

  def slow_function() do
    Process.sleep(3000)
    42
  end

  def supervisor() do
    pid = spawn(fn -> spin("thinking!") end)
    result = slow_function()
    send(pid, {self(), :computed})

    receive do
      :done -> result
    end
  end

  def main() do
    result = supervisor()
    IO.puts("Answer: #{result}")
  end
end

Spinner.main()
